pandapipes/component_models/abstract_models/circulation_pump.py

```diff
@@ -43,30 +113,34 @@ def extract_results(cls, net, options, branch_results, nodes_connected, branches
         :type options:
         :return: No Output.
         """
-        res_table, circ_pump, index_nodes_from, node_pit, _ = \
-            super().extract_results(net, options, None, nodes_connected, branches_connected)
+        circ_pump_tbl = net[cls.table_name()]
 
-        index_juncts_to = circ_pump.to_junction.values
-        junct_uni_to = np.array(list(set(index_juncts_to)))
-        index_nodes_to = get_lookup(net, "node", "index")[
-            cls.get_connected_node_type().table_name()][junct_uni_to]
+        if len(circ_pump_tbl) == 0:
+            return
```

Comment:
There have been a lot of changes to this function, which is why it's not so easy to follow up.
1. Before, circulation pump inherited from ext_grid, and the largest part of the result extraction was done there. Now, it is all implemented in the circulation pump itself.
2. One functionality has been externalized (`get_mass_flow_at_nodes`), as it is used by different components. It performs a mass flow balance around fixed p nodes to identify the missing flow.
Line 118/119 just state that in case of an empty table, no results shall be extracted to avoid errors to be raised just due to facts like elements being dropped in the pandas table, without deleting the component from the component list.